### PR TITLE
Remove panics; replace with error (FF-2512)

### PR DESCRIPTION
**Pull request by leoromanovsky**
## motivation

Several users have given us feedback that the behavior of our SDK to issue panics is highly detrimental to their servers. Unlike other languages like Swift which have a `throws` type, we are unable to warn callers with compiler types to expect a function to issue a panic. 

In golang a panic is commonly issued when we experience an error type that we are unable to handle. If the caller does not explicitly handle the panic, their program will exit with a non-zero code.

https://gobyexample.com/panic

## changes

Removes all usages of `panic` from the codebase and replaces them with `error`. 

The usages below are for handling panics from the assignment logger.

```
➜ grep -r 'panic' .
./eppoclient/lruassignmentlogger_test.go:func Test_LruAssignmentLogger_panicsAreNotCached(t *testing.T) {
./eppoclient/lruassignmentlogger_test.go:	innerLogger.On("LogAssignment", mock.Anything).Panic("test panic")
./eppoclient/lruassignmentlogger.go:		// case it panics.
./eppoclient/client.go:			// need to catch panics from Logger and continue
./eppoclient/client.go:					fmt.Println("panic occurred:", r)
./eppoclient/eppoclient_e2e_test.go:						panic(fmt.Sprintf("unknown variation: %v", test.VariationType))
./eppoclient/eppoclient_e2e_test.go:		panic("test cases files read error")
./eppoclient/poller_test.go:			panic("some_error")
./eppoclient/client_test.go:	mockLogger.Mock.On("LogAssignment", mock.Anything).Panic("logging panic")
```

This is a canonical mechanism in go to return the desired value and an error as a tuple, such as `(boolean, error)`.

When callers encounter a tuple like this they are expected to first check for the `error` to be non-null, such as:

```
isFound, err := myFunc(...)
if err != nil {
  // handle: return, panic, log - the callers choice
}
```

By returning an error we use compiler types to warn the users that the function might return exceptional values and they need to be defensive. It is common, and not necessarily an elegant part of the go language, to see widespread use of error handling.

## other changes

1. the SDK previously panic on all `getAssignment` calls when the last configuration request failed with a http 403. This is in theory a useful mechanism to disable to the SDK in the case of invalid SDK key. However there can be innocent reasons for this status code outside the users maleficence. Furthermore throwing a panic is an extremely aggressive and disruptive response. Eppo will develop other mechanisms to enforcing valid licensing.
2. Correcting documentation of constructor. The only exported and possible way to create a client is with the `InitClient` method.
3. Documented assignment cache.

## for reviewers

* The `matches` method only returns a `bool` and at present swallows the error generated by invalid conditions lower down. Should I pass it all the way up? My hunch is yes but I wonder if its actionable by the user. At worst it is useful debugging information to pass to Eppo support, but it can also be logged to `stdout`

### CODE REVIEW

File: README.md
Comment (leoromanovsky) [inline README.md:46]: This code example was not valid - the `EppoClient` cannot be instantiated in the way it was documented. Instead we create a global variable and declare its type; it is currently `nil`.

File: README.md
Comment (leoromanovsky) [inline README.md:57]: Allow the caller to handle failures to initialize as they desire - it is obvious from the compiler type that an `error` is returned.

👉  We could consider this a breaking API change and publish a v5.

File: README.md
Comment (leoromanovsky) [inline README.md:136]: This documentation was missing; the feature has already been released.

File: eppoclient/client.go
Comment (leoromanovsky) [inline eppoclient/client.go:18]: making it clear that this is not available publically

File: eppoclient/client.go
Comment (leoromanovsky) [inline eppoclient/client.go:91]: while this is extremely undesirable inputs; they will return the default value and the user can handle the error safely. the panic made sense during previous implementations of the SDK before we added default values.

File: eppoclient/client_test.go
Comment (leoromanovsky) [inline eppoclient/client_test.go:17]: test matches new behavior

File: eppoclient/configurationrequestor.go
Comment (leoromanovsky) [inline eppoclient/configurationrequestor.go:31]: removing this behavior entirely

File: eppoclient/lruassignmentlogger.go
Comment (leoromanovsky) [inline eppoclient/lruassignmentlogger.go:38]: the linter reported that `self` or `this` was not desirable and that the variable should be more descriptive; I have used a shorthand for the struct previously.

File: eppoclient/rules.go
Comment (leoromanovsky) [inline eppoclient/rules.go:57]: @greghuels What is most helpful to add foundationally here to support evaluation reasons? I am thinking changing the function definition to `(bool, error)` so we can give the highest fidelity error up to the `getAssignment` method about the nature of the failure.

File: eppoclient/rules.go
Comment (leoromanovsky) [inline eppoclient/rules.go:169]: this isn't pretty but is common for golang code

**Comment by leoromanovsky** [comment]
@evgendn We are reviewing removing panics from the golang SDK - thank you for your feedback earlier.

File: eppoclient/client.go
Comment (rasendubi) [reply eppoclient/client.go:18]: nit: that's a redundant comment in Go as everything starting with lowercase letter is not exported

File: eppoclient/configurationrequestor.go
Comment (rasendubi) [reply eppoclient/configurationrequestor.go:31]: This will conflict with https://github.com/Eppo-exp/golang-sdk/pull/50. I can rebase one PR on top of another, depending on which we merge first.

File: eppoclient/lruassignmentlogger.go
Comment (rasendubi) [reply eppoclient/lruassignmentlogger.go:38]: Coming from languages that enforce `self`, that's pretty much unfortunate. I don't think `lal` is any more descriptive than `self` here.

File: eppoclient/rules.go
Comment (rasendubi) [inline eppoclient/rules.go:159]: minor: `promotedValue == value` suggest that they are the same value. I would rather rename `promotedValue` to `promotedAttributeValue` or similar.

File: eppoclient/rules.go
Comment (rasendubi) [inline eppoclient/rules.go:182]: minor: `evaluateSemVerCondition`, `evaluateNumericCondition`, `promoteInt`, `promoteUint`, `promoteFloat` are internal functions and they expect arguments to be of specific type (and they are properly guarded in our code). There's no legitimate reason for them to return an error, so panic seems appropriate as it signifies a programming error.

File: eppoclient/rules.go
Comment (rasendubi) [reply eppoclient/rules.go:57]: See my other comment: `evaluateNumericCondition` has no legitimate reason to return an error. An error here means we have a bug in SDK.

Comment (felipecsl) [reply eppoclient/rules.go:57]: +1 to changing the signature here to return `(bool, error)` - we don't want to just silently swallow it

File: eppoclient/rules.go
Comment (felipecsl) [inline eppoclient/rules.go:76]: same here

**Review by felipecsl — COMMENTED** [review]
Main feedback around returning errors instead of swallowing it, otherwise great improvement, we were certainly too panic-happy before

**Comment by guppykang** [comment]
thank you 🙏

File: README.md
Comment (aarsilv) [reply README.md:57]: Yeah with this signature being different it's probably v5

File: README.md
Comment (aarsilv) [reply README.md:136]: 🙌

File: eppoclient/lruassignmentlogger.go
Comment (aarsilv) [reply eppoclient/lruassignmentlogger.go:38]: how about `logger` or `assignmentLogger`?

File: eppoclient/rules.go
Comment (aarsilv) [reply eppoclient/rules.go:57]: I do think we should surface _unexpected_ errors the SDK encounters. So I side with the above two comments.

Anticipated error conditions, like the user passing a string-valued attribute for comparison using a numeric operator, should return `false` like we do.

Side note: it looks like in JavaScript, if the attribute isn't a number, we just compare the non-number as a string to the target number as a string 😬 I can file a bug for that.

File: eppoclient/rules.go
Comment (aarsilv) [reply eppoclient/rules.go:182]: But we don't want bugs / programming errors on our part to tank the users application--so thinking we should avoid `panics` all together and instead our errors will surface as a top-levle error being returned by the `getAssignment()` methods

**Review by aarsilv — APPROVED** [review]
Thanks for digging into this! Approving assuming we return the error along with the boolean for numeric comparisons.

File: eppoclient/rules.go
Comment (rasendubi) [reply eppoclient/rules.go:57]: There are a couple of other options to achieve this:
- Inline these functions in the call site. The call site has the proper guards, so error branch becomes unreachable—there will be no panics nor errors.
- Wrap the whole evaluation into a recover, so the panic does not reach the user. It's probably a good idea to do this anyway—there are many more ways Go can panic without an explicit `panic` call

File: eppoclient/rules.go
Comment (rasendubi) [reply eppoclient/rules.go:182]: Let's merge this discussion with https://github.com/Eppo-exp/golang-sdk/pull/54#discussion_r1656756419

File: eppoclient/configurationrequestor.go
Comment (leoromanovsky) [reply eppoclient/configurationrequestor.go:31]: I will wait for your PR to merge since it is larger in scope.

File: eppoclient/lruassignmentlogger.go
Comment (leoromanovsky) [reply eppoclient/lruassignmentlogger.go:38]: It is common to name the `method received` with one character, the lowercase version of the first character in the implementing struct.

https://go.dev/wiki/CodeReviewComments#receiver-names

```
Receiver Names

The name of a method’s receiver should be a reflection of its identity; often a one or two letter abbreviation of its type suffices (such as “c” or “cl” for “Client”). Don’t use generic names such as “me”, “this” or “self”, identifiers typical of object-oriented languages that gives the method a special meaning. In Go, the receiver of a method is just another parameter and therefore, should be named accordingly. The name need not be as descriptive as that of a method argument, as its role is obvious and serves no documentary purpose. It can be very short as it will appear on almost every line of every method of the type; familiarity admits brevity. Be consistent, too: if you call the receiver “c” in one method, don’t call it “cl” in another.
```

File: eppoclient/rules.go
Comment (leoromanovsky) [reply eppoclient/rules.go:57]: > Wrap the whole evaluation into a recover, so the panic does not reach the user. It's probably a good idea to do this anyway because we can't guard against all potential panics

Yea this makes sense to me - will catch the panic and return default value. Great suggestion.

File: README.md
Comment (felipecsl) [reply README.md:57]: +1 to publishing a new major version according to semver

**Review by felipecsl — APPROVED** [review]

File: eppoclient/client.go
Comment (leoromanovsky) [reply eppoclient/client.go:18]: You're right - thank you.

**Comment by leoromanovsky** [comment]
Getting familiar with golang code from reading some libraries: I don't think ALL functions need to return an `error` - sometimes they can handle it locally without making it another functions problem. However in the interest of supporting "detailed evaluation reasons" that is forthcoming, the rule evaluation functions should return an error so that we can provide a high fidelity rationale for why an allocation did not match.

**Pull request merged**
